### PR TITLE
Migrations Backups

### PR DESCRIPTION
**Pull request by fogelito**
- Appwrite refactors for backup migrations

### CODE REVIEW

File: src/Appwrite/Platform/Tasks/Maintenance.php [hunk 1]
```diff
@@ -64,6 +59,21 @@ public function action(Database $dbForConsole, Certificate $queueForCertificates
         }, $interval, $delay);
     }
 
+    /**
+     * Overridable method for Cloud, Please do not delete
+     */
+    protected function notifyProjects(Delete $queueForDeletes, int $usageStatsRetentionHourly): void
+    {
+        var_dump("Appwrite notifyProjects");
+
```
Comment (abnegate) [inline src/Appwrite/Platform/Tasks/Maintenance.php:68]:
```suggestion
```

File: src/Appwrite/Platform/Tasks/Maintenance.php [hunk 2]
```diff
@@ -64,6 +59,21 @@ public function action(Database $dbForConsole, Certificate $queueForCertificates
         }, $interval, $delay);
     }
 
+    /**
+     * Overridable method for Cloud, Please do not delete
```
Comment (abnegate) [inline src/Appwrite/Platform/Tasks/Maintenance.php:63]:
```suggestion
     * Hook to allow sub-classes to extend project-level maintenance functionality.
```

File: src/Appwrite/Platform/Tasks/ScheduleBase.php [hunk 3]
```diff
@@ -62,13 +59,12 @@ public function action(Group $pools, Database $dbForConsole, callable $getProjec
         $getSchedule = function (Document $schedule) use ($dbForConsole, $getProjectDB): array {
             $project = $dbForConsole->getDocument('projects', $schedule->getAttribute('projectId'));
 
-            $collectionId = match ($schedule->getAttribute('resourceType')) {
-                'function' => 'functions',
-                'message' => 'messages'
-            };
+            var_dump('===== $getSchedule getCollectionId = ');
+            var_dump(static::getCollectionId());
+            var_dump('===== $getSchedule getCollectionId = ');
```
Comment (abnegate) [inline src/Appwrite/Platform/Tasks/ScheduleBase.php:64]:
```suggestion
```

File: src/Appwrite/Platform/Tasks/ScheduleBase.php [hunk 4]
```diff
@@ -109,13 +105,13 @@ public function action(Group $pools, Database $dbForConsole, callable $getProjec
 
             foreach ($results as $document) {
                 try {
+                    var_dump('=== ScheduleBase start');
+                    var_dump($getSchedule($document)['resource']);
+                    var_dump('=== ScheduleBase end');
+                    //todo: use a unique key as InternalId or add projectId
```
Comment (abnegate) [inline src/Appwrite/Platform/Tasks/ScheduleBase.php:111]:
```suggestion
```

File: src/Appwrite/Platform/Tasks/ScheduleFunctions.php [hunk 5]
```diff
@@ -26,7 +26,12 @@ public static function getSupportedResource(): string
         return 'function';
     }
 
-    protected function enqueueResources(Group $pools, Database $dbForConsole): void
+    public static function getCollectionId(): string
```
Comment (abnegate) [inline src/Appwrite/Platform/Tasks/ScheduleFunctions.php:29]: Nice, this a much better approach than the switch :+1:

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 6]
```diff
@@ -4783,4 +4785,56 @@ public function testTimeout(array $data): void
 
         $this->assertEquals(408, $response['headers']['status-code']);
     }
+
+    /**
+     * @depends testCreateDatabase
+     */
+    public function testPreserveDates(array $data): void
```
Comment (abnegate) [inline tests/e2e/Services/Databases/DatabasesBase.php:4792]: We can remove this now

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 7]
```diff
@@ -4754,14 +4754,16 @@ public function testTimeout(array $data): void
...
+
+        for ($i = 0; $i < 1; $i++) {
             $this->client->call(Client::METHOD_POST, '/databases/' . $data['databaseId'] . '/collections/' . $data['$id'] . '/documents', array_merge([
                 'content-type' => 'application/json',
                 'x-appwrite-project' => $this->getProject()['$id'],
             ], $this->getHeaders()), [
                 'documentId' => ID::unique(),
                 'data' => [
-                    'longtext' => file_get_contents(__DIR__ . '/../../../resources/longtext.txt'),
+                    'longtext' => $longtext . $longtext,
```
Comment (abnegate) [inline tests/e2e/Services/Databases/DatabasesBase.php:4766]: Do we need these changes for this PR?

File: src/Appwrite/Platform/Tasks/Maintenance.php (hunk 1 above)
Comment (fogelito) [reply src/Appwrite/Platform/Tasks/Maintenance.php:68]: Done!

File: src/Appwrite/Platform/Tasks/ScheduleBase.php (hunk 3 above)
Comment (fogelito) [reply src/Appwrite/Platform/Tasks/ScheduleBase.php:64]: done!

File: src/Appwrite/Platform/Tasks/ScheduleBase.php (hunk 4 above)
Comment (fogelito) [reply src/Appwrite/Platform/Tasks/ScheduleBase.php:111]: Removed the var_dumps added a remonder to add 'projectId' value

File: tests/e2e/Services/Databases/DatabasesBase.php (hunk 6 above)
Comment (fogelito) [reply tests/e2e/Services/Databases/DatabasesBase.php:4792]: Done!

File: tests/e2e/Services/Databases/DatabasesBase.php (hunk 7 above)
Comment (fogelito) [reply tests/e2e/Services/Databases/DatabasesBase.php:4766]: Nop.. Reverted

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 8]
```diff
@@ -4754,7 +4754,7 @@ public function testTimeout(array $data): void
 
         $this->assertEquals($longtext['headers']['status-code'], 202);
 
-        for ($i = 0; $i < 10; $i++) {
+        for ($i = 0; $i < 1; $i++) {
```
Comment (abnegate) [inline tests/e2e/Services/Databases/DatabasesBase.php:4757]: We need this as 10 for cloud tests, MySQL never times out with only 1

Comment (fogelito) [reply tests/e2e/Services/Databases/DatabasesBase.php:4757]: Done!

File: app/config/scopes.php [hunk 9]
```diff
@@ -130,4 +130,22 @@
...
+    ],
+    'archives.write' => [
+        'description' => 'Access to create, update, and delete your archive',
+    ],
+    'restorations.read' => [
+        'description' => 'Access to read your restorations',
+    ],
+    'restorations.write' => [
+        'description' => 'Access to create, update, and delete your restoration',
+    ],
```
[the remark this answers is not in the document]
Comment (fogelito) [reply app/config/scopes.php]: Done!

File: app/config/platforms.php [hunk 10]
```diff
@@ -1,9 +1,5 @@
 <?php
 
-const APP_PLATFORM_SERVER = 'server';
-const APP_PLATFORM_CLIENT = 'client';
-const APP_PLATFORM_CONSOLE = 'console';
```
Comment (christyjacob4) [inline app/config/platforms.php:5]: why was this removed?

File: src/Appwrite/Specification/Format/Swagger2.php [hunk 11]
```diff
@@ -286,13 +286,26 @@ public function parse(): array
...
+
+                $base = !empty($class)
+                    ? \get_parent_class($class)
+                    : '';
+
+                switch ($base) {
+                    case 'Appwrite\Utopia\Database\Validator\Queries\Base':
+                        $class = $base;
+                        break;
+                }
```
Comment (christyjacob4) [inline src/Appwrite/Specification/Format/Swagger2.php:301]: can you run docker compose exec appwrite specs for both the latest and 1.6.x versions so we can see if this change has affected the specs in any way?

File: app/config/collections.php [hunk 12]
```diff
@@ -4138,6 +4149,28 @@
...
+                '$id' => ID::custom('resourceType'),
+                'type' => Database::VAR_STRING,
+                'format' => '',
+                'size' => Database::LENGTH_KEY,
+                'signed' => true,
+                'required' => false,
+                'default' => null,
+                'array' => false,
+                'filters' => [],
+            ],
```
Comment (christyjacob4) [inline app/config/collections.php:4173]: don't we need any new indexes?

File: app/config/platforms.php (hunk 10 above)
Comment (fogelito) [reply app/config/platforms.php]: Reverted 👍

File: src/Appwrite/Platform/Workers/Migrations.php [hunk 13]
```diff
@@ -233,99 +254,120 @@ protected function generateAPIKey(Document $project): Document
...
+            }
 
-            $destination = new DestinationsAppwrite(
-                $projectDocument->getId(),
-                'http://appwrite/v1',
-                $tempAPIKey['secret'],
-            );
+            $migration->setAttribute('stage', 'processing');
+            $migration->setAttribute('status', 'processing');
+            $log->addBreadcrumb(new Breadcrumb('debug', 'migration', "Migration hit stage 'processing'", \microtime(true)));
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Migrations.php:290]: I think breadcrumbs were removed and not supported anymore. Please check

File: src/Appwrite/Platform/Workers/Migrations.php [hunk 14]
```diff
@@ -337,26 +379,28 @@ protected function processMigration(Document $project, Document $migration, Log
...
             }
-            if ($migrationDocument) {
-                $this->updateMigrationDocument($migrationDocument, $projectDocument);
 
-                if ($migrationDocument->getAttribute('status', '') == 'failed') {
-                    throw new Exception("Migration failed");
-                }
+            $this->updateMigrationDocument($migration, $projectDocument);
+
+            if ($migration->getAttribute('status', '') == 'failed') {
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Migrations.php:399]:
```suggestion
            if ($migration->getAttribute('status', '') === 'failed') {
```

File: src/Appwrite/Platform/Workers/Migrations.php [hunk 15]
```diff
@@ -337,26 +379,28 @@ protected function processMigration(Document $project, Document $migration, Log
...
+                $destination->error();
+                $source->error();
+
+                throw new Exception('Migration failed');
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Migrations.php:403]: Can we also report the reason for the "Migration Failed" error? Maybe convert the error message to a string and attach it to the exception?

File: app/config/collections.php (hunk 12 above)
Comment (fogelito) [reply app/config/collections.php:4173]: We do not search in the migration collection by resourceId resourceType it is only a carrier of this data.

File: src/Appwrite/Platform/Workers/Migrations.php (hunk 15 above)
Comment (fogelito) [reply src/Appwrite/Platform/Workers/Migrations.php:403]: The errors of sources and destinations are appended, not always have a single error, sometimes the errors list can be very big. I guess we can always make the errors mechanism better in Migrations.php

File: src/Appwrite/Platform/Workers/Migrations.php (hunk 13 above)
Comment (fogelito) [reply src/Appwrite/Platform/Workers/Migrations.php]: Seems like no one is using it 👍 
Removed

File: src/Appwrite/Specification/Format/Swagger2.php (hunk 11 above)
Comment (fogelito) [reply src/Appwrite/Specification/Format/Swagger2.php:301]: I see $createdAt $updatedAt were added , that is expected since I added them for attributes & collections?

File: src/Appwrite/Platform/Workers/Migrations.php (hunk 15 above)
Comment (christyjacob4) [reply src/Appwrite/Platform/Workers/Migrations.php:399]: okay 👍

**Review by christyjacob4 — APPROVED** [review]

File: app/controllers/api/account.php [hunk 16]
```diff
@@ -3683,7 +3683,7 @@ function sendSessionAlert(Locale $locale, Document $user, Document $project, Doc
     });
 
 App::get('/v1/account/mfa/factors')
-    ->desc('List Factors')
```
Comment (TorstenDittmann) [inline app/controllers/api/account.php:3686]: what are all these lower case changes for?

they dont work good in avatars for example

File: app/controllers/api/teams.php [hunk 17]
```diff
@@ -868,7 +879,17 @@
     ->label('sdk.response.model', Response::MODEL_MEMBERSHIP)
     ->param('teamId', '', new UID(), 'Team ID.')
     ->param('membershipId', '', new UID(), 'Membership ID.')
-    ->param('roles', [], new ArrayList(new Key(), APP_LIMIT_ARRAY_PARAMS_SIZE), 'An array of strings. Use this param to set the user\'s roles in the team. A role can be any string. Learn more about [roles and permissions](https://appwrite.io/docs/permissions). Maximum of ' . APP_LIMIT_ARRAY_PARAMS_SIZE . ' roles are allowed, each 32 characters long.')
+    ->param('roles', [], function (Document $project) {
+        if ($project->getId() === 'console') {
+            ;
```
Comment (TorstenDittmann) [inline app/controllers/api/teams.php:884]: a lonely semi-colon?

File: app/controllers/api/teams.php [hunk 18]
```diff
@@ -868,7 +879,17 @@
...
+            $roles = array_keys(Config::getParam('roles', []));
```
Comment (TorstenDittmann) [inline app/controllers/api/teams.php:885]: also these role changes should not be part of this PR right?

File: app/init.php [hunk 19]
```diff
@@ -1258,13 +1262,13 @@ function (mixed $value) {
...
-            $user = new Document([]);
-        }
-    }
+    // if (APP_MODE_ADMIN === $mode) {
+    //     if ($user->find('teamInternalId', $project->getAttribute('teamInternalId'), 'memberships')) {
+    //         Authorization::setDefaultStatus(false);  // Cancel security segmentation for admin users.
+    //     } else {
+    //         $user = new Document([]);
+    //     }
+    // }
```
Comment (TorstenDittmann) [inline app/init.php:1271]: is this not needed?

**Review by TorstenDittmann — APPROVED** [review]
